chef.py

```diff
@@ -0,0 +1,459 @@
+
+        # Polyfill localStorage and document.cookie as iframes can't access
+        # them
+        return (content
+            .replace("localStorage", "_localStorage")
+            .replace('document.cookie.split', '"".split')
+            .replace('document.cookie', 'window._document_cookie'))
+
+    def css_node_filter(node):
+        return "stylesheet" in node["rel"]
+
+    def css_content_middleware(content, url, **kwargs):
+        file_dir = os.path.dirname(urlparse(url).path)
+
+        # Download linked fonts and images
+        def repl(match):
+            src = match.group(1)
+            if src.startswith('//localhost'):
+                return 'url()'
+            # Don't download data: files
+            if src.startswith('data:'):
+                return match.group(0)
+            src_url = make_fully_qualified_url(os.path.join(file_dir, src))
+            derived_filename = derive_filename(src_url)
+            download_file(src_url, destination, request_fn=make_request,
+                    filename=derived_filename)
+            return 'url("%s")' % derived_filename
+
+        return CSS_URL_RE.sub(repl, content)
+
+    # Download all linked static assets.
+    download_assets("img[src]", "src")  # Images
+    download_assets("link[href]", "href",
+            content_middleware=css_content_middleware,
+            node_filter=css_node_filter)  # CSS
+    download_assets("script[src]", "src", content_middleware=js_middleware) # JS
+    download_assets("source[src]", "src") # Audio
+    download_assets("source[srcset]", "srcset") # Audio
+
+    # ... and also run the middleware on CSS/JS embedded in the page source to
+    # get linked files.
+    for node in doc.select('style'):
+        node.string = css_content_middleware(node.get_text(), url='')
+
+    for node in doc.select('script'):
+        if not node.attrs.get('src'):
+            node.string = js_middleware(node.get_text(), url='')
+
+    return doc
+
+
+url_blacklist = [
+    'analytics.js',
+]
+
+def is_blacklisted(url):
+    return any((item in url) for item in url_blacklist)
+
+
+def derive_filename(url):
+    return "%s.%s" % (uuid.uuid4().hex, os.path.basename(urlparse(url).path))
+
+
+def make_request(url, clear_cookies=True, timeout=60, *args, **kwargs):
+    if clear_cookies:
+        sess.cookies.clear()
+
+    retry_count = 0
+    max_retries = 5
+    while True:
+        try:
+            response = sess.get(url, headers=headers, timeout=timeout, *args, **kwargs)
+            break
+        except (requests.exceptions.ConnectionError, requests.exceptions.ReadTimeout) as e:
+            retry_count += 1
+            print("Error with connection ('{msg}'); about to perform retry {count} of {trymax}."
+                  .format(msg=str(e), count=retry_count, trymax=max_retries))
+            time.sleep(retry_count * 1)
+            if retry_count >= max_retries:
+                return Dummy404ResponseObject(url=url)
```

Comment:
If you want this not to be dangling, here it is copy-pasted from https://github.com/learningequality/sushi-chef-ck12/blob/master/chef.py#L77

```
class Dummy404ResponseObject(requests.Response):
    def __init__(self, url):
        super(Dummy404ResponseObject, self).__init__()
        self._content = b""
        self.status_code = 404
        self.url = url
```